File: __tests__/unit/model/hand/get/getStraightFlush.ts
import getStraightFlush from "../../../../../src/model/hand/get/getStraightFlush";
import cards from "../../../../../src/model/card/cards";
import { HandType } from "../../../../../src/model/hand/Hand";
import { Value } from "../../../../../src/model/card/Card";

describe("retrieve straight flushes", () => {
  test("retrieve a straight flush", () => {
    const straightFlush = getStraightFlush(cards("2h", "3h", "4h", "5h", "6h"));
    expect(straightFlush.type).toEqual(HandType.StraightFlush);
  });

  test("retrieve a straight flush from ace to 5", () => {
    const straightFlush = getStraightFlush(cards("Ah", "2h", "3h", "4h", "5h"));
    expect(straightFlush.type).toEqual(HandType.StraightFlush);
  });

  test("retrieves the highest straight flush", () => {
    const straightFlush = getStraightFlush(
      cards("Ah", "2h", "3h", "4h", "5h", "6h", "7h")
    );
    expect(straightFlush.type).toEqual(HandType.StraightFlush);
    expect(straightFlush.value()).toEqual(Value.seven);
  });
});

File: src/rust/src/tests/get_straight_flush.rs
use crate::card::{Card, Suit, Value};
use crate::hand::Hand;

#[test]
fn test_get_straight_flush() {
    // Test with a straight flush
    let cards = vec![
        Card::new(Suit::Hearts, Value::Two),
        Card::new(Suit::Hearts, Value::Three),
        Card::new(Suit::Hearts, Value::Four),
        Card::new(Suit::Hearts, Value::Five),
        Card::new(Suit::Hearts, Value::Six),
        // Add some other cards
        Card::new(Suit::Diamonds, Value::Seven),
        Card::new(Suit::Clubs, Value::Eight),
    ];

    let hand = Hand::new(cards);
    let straight_flush = hand.get_straight_flush();
    assert_eq!(straight_flush.len(), 5);
    assert_eq!(straight_flush[0].suit(), Suit::Hearts);
    assert_eq!(straight_flush[0].value(), Value::Two);
    assert_eq!(straight_flush[1].value(), Value::Three);
    assert_eq!(straight_flush[2].value(), Value::Four);
    assert_eq!(straight_flush[3].value(), Value::Five);
    assert_eq!(straight_flush[4].value(), Value::Six);

    // Test without a straight flush
    let cards = vec![
        Card::new(Suit::Hearts, Value::Two),
        Card::new(Suit::Hearts, Value::Three),
        Card::new(Suit::Hearts, Value::Four),
        Card::new(Suit::Hearts, Value::Five),
        Card::new(Suit::Diamonds, Value::Six),
        Card::new(Suit::Clubs, Value::Seven),
    ];

    let hand = Hand::new(cards);
    let straight_flush = hand.get_straight_flush();
    assert!(straight_flush.is_empty());

    // Test with a straight but not a flush
    let cards = vec![
        Card::new(Suit::Hearts, Value::Two),
        Card::new(Suit::Diamonds, Value::Three),
        Card::new(Suit::Clubs, Value::Four),
        Card::new(Suit::Spades, Value::Five),
        Card::new(Suit::Hearts, Value::Six),
        Card::new(Suit::Clubs, Value::Seven),
    ];

    let hand = Hand::new(cards);
    let straight_flush = hand.get_straight_flush();
    assert!(straight_flush.is_empty());

    // Test with a flush but not a straight
    let cards = vec![
        Card::new(Suit::Hearts, Value::Two),
        Card::new(Suit::Hearts, Value::Four),
        Card::new(Suit::Hearts, Value::Six),
        Card::new(Suit::Hearts, Value::Eight),
        Card::new(Suit::Hearts, Value::Ten),
        Card::new(Suit::Clubs, Value::Seven),
    ];

    let hand = Hand::new(cards);
    let straight_flush = hand.get_straight_flush();
    assert!(straight_flush.is_empty());
}
